std/functional.d

```diff
@@ -899,6 +903,149 @@ template partial(alias fun, alias arg)
     assert(result2 == 1.5f);
 }
 
+/**
+Takes a function of (potentially) many arguments, and returns a function taking
+one argument and returns a callable taking the rest.  f(x, y) == curry(f)(x)(y)
+
+Params:
+    F = a function taking at least one argument
+    t = a callable object whose opCall takes at least 1 object
+Returns:
+    A single parameter callable object
+*/
+template curry(alias F)
```

Comment:
`if 'isCallable!F` from `std.traits` should work as a constraint.

Reply:
I've seen only later the other overload (for opCall) so maybe no possible

Reply:
I wrapped the first overload in static if(isCallable!F) and my tests still work.  The important distinction is "is there state to be stored or not"

Reply:
`isCallable` should definitly be a template constraint to avoid akward template errors, e.g.:

```
curry!1();

std\functional.d(950): Error: function expected before (), not curry!1 of type void
    curry!1();
```